goose.go

```diff
@@ -2045,8 +2109,67 @@ func (ctx Ctx) imports(d []ast.Spec) []coq.Decl {
 func (ctx Ctx) maybeDecls(d ast.Decl) []coq.Decl {
 	switch d := d.(type) {
 	case *ast.FuncDecl:
+		cv := coq.StructToInterface{}
+		for i := 0; i < len(d.Body.List); i++ {
+			switch f := d.Body.List[i].(type) {
+			case *ast.ReturnStmt:
+				interfaceName := ""
+				methods := []string{}
+				if len(f.Results) > 0 {
+					if results, ok := f.Results[0].(*ast.BinaryExpr); ok {
+						if call, ok := results.X.(*ast.CallExpr); ok {
+							if signature, ok := ctx.typeOf(call.Fun).(*types.Signature); ok {
```

Comment:
Oh dear, this is getting extremely nested. I think there are a couple things to do about this:

- You should break out as much as you can to helper functions. For example, at least the `case *ast.ReturnStmt` and `case *ExprStmt` should be separate functions (I'm not exactly sure what these two cases do). There's also a pattern that gets and cleans up type names.
- These `if` statements that do type switching necessarily leads to one level of nesting, which is annoying. What you can try to do is factor out a function which does all the type-checking and if it succeeds returns the `*ast.CallExpr` or whatever else you're looking for.